Add transitions to components

capsula4 opened the pull request:
Apply small transition for a smoother hover effect on inputs/buttons.

Depends on https://github.com/bigcommerce/big-design/pull/273

chanceaclark commented:
Hey @capsula4, adding transitions/animations is definitely an action item. If we are to add them, we should probably add this as a full fledge feature instead and get it approved through design. What do you think?

capsula4 commented:
@chanceaclark sure, will talk with Valter but I guess some transitions are better than no transitions at all. IMO we don't need to tackle everything at once.

capsula4 commented:
@chanceaclark  as per discussed, hard-coded the transitions for now, until we figure out how we want to expose this properly via the theme

capsula4 commented:
Note to self: only transition the specific properties

chanceaclark commented on packages/big-design/src/components/Button/styled.tsx:
```diff
@@ -30,6 +30,7 @@ export const StyledButton = styled.button<ButtonProps & MarginProps>`
   position: relative;
   text-align: center;
   text-decoration: none;
+  transition: background-color 150ms ease-out;
```
For active, focus, and disabled state:
```suggestion
  transition: 150ms ease-out;
  transition-property: background-color, border-color, box-shadow, color;
```

chanceaclark commented on packages/big-design/src/components/Link/styled.tsx:
```diff
@@ -13,6 +13,15 @@ export const StyledLink = styled.a<LinkProps & { isExternal?: boolean }>`
+
+  &:hover:not(:active) {
+    background-color: ${({ theme }) => theme.colors.primary50};
+  }
```
Did you intend for this to be `color` instead of `background-color`?

chanceaclark commented on packages/big-design/src/components/Checkbox/styled.tsx:
```diff
@@ -32,25 +32,32 @@ export const StyledCheckbox = styled.label<StyledCheckboxProps>`
   height: ${({ theme }) => theme.spacing.large};
   justify-content: center;
-  transition: all 150ms;
+  transition: all 150ms ease-out;
```
I know this was done before this PR, but could you explicitly set the transition instead of using `all`? 
```suggestion
  transition: 150ms ease-out;
  transition-property: border-color, box-shadow, color, visibility;
```

chanceaclark commented on packages/big-design/src/components/Checkbox/styled.tsx:
```diff
@@ -32,25 +32,32 @@ export const StyledCheckbox = styled.label<StyledCheckboxProps>`
+    border-color: ${({ checked, isIndeterminate, theme }) =>
+      checked || isIndeterminate ? theme.colors.primary : theme.colors.secondary40};
+  }
+
```
This doesn't work well with `disabled` checkboxes. Can you account for this?

capsula4 replied:
I have fixed it, however disabled checkboxes seem to be broken at the moment. You can still click on disabled checkboxes.

chanceaclark commented on packages/big-design/src/components/List/Item/styled.tsx:
```diff
@@ -13,6 +13,8 @@ export const StyledListItem = styled.li<ListItemProps>`
   min-width: ${({ theme }) => theme.helpers.remCalc(256)};
   outline: none;
   padding: 0 ${({ theme }) => theme.spacing.medium};
+  transition: all 150ms ease-out;
```
Shouldn't need all if using `transition-property`
```suggestion
  transition: 150ms ease-out;
```

chanceaclark commented on packages/big-design/src/components/ProgressBar/styled.tsx:
```diff
@@ -18,7 +18,7 @@ export const StyledProgressBarFiller = styled.div<ProgressBarProps>`
     typeof percent === 'number'
       ? css`
-          transition: width 0.2s ease-in;
+          transition: all 150ms ease-out;
```
Instead of `all`:
```suggestion
          transition: 150ms ease-out;
          transition-property: left, width;
```

chanceaclark commented on packages/big-design/src/components/Radio/styled.tsx:
```diff
@@ -28,30 +29,41 @@ export const StyledRadio = styled.label<StyledRadioProps>`
 
+  &:hover {
+    border-color: ${props => (props.checked ? props.theme.colors.primary50 : props.theme.colors.secondary40)};
+  }
```
Same issue with checkboxes, doesn't work well with `disabled`.

capsula4 replied:
at the moment there doesn't seem to be support for disabled radio buttons (no styling)

chanceaclark commented on packages/big-design/src/components/Textarea/styled.tsx:
```diff
@@ -17,6 +17,7 @@ export const StyledTextarea = styled.textarea<TextareaProps>`
   line-height: ${({ theme }) => theme.lineHeight.medium};
   max-height: ${({ theme }) => theme.helpers.remCalc(224)};
   padding: ${({ theme }) => `${theme.spacing.xxSmall} ${theme.spacing.small}`};
+  transition: border 150ms ease-out;
```
Focus state:
```suggestion
  transition: 150ms ease-out;
  transition-property: border, box-shadow;
```

capsula4 commented:
@chanceaclark thanks for the thorough review. I have addressed most of your comments, and created a mixing so things are not too hard-coded. Eventually the mixin could consume the values from the theme. 

Disabled state for checkbox is broken at the moment, maybe better to fix in a different PR.
Radio buttons do not have support for disabled state.

chanceaclark commented on packages/big-design/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -26,7 +26,7 @@ class RawCheckbox extends React.PureComponent<CheckboxProps & PrivateProps> {
 
   render() {
-    const { checked, className, isIndeterminate, label, forwardedRef, style, ...props } = this.props;
+    const { checked, className, disabled, isIndeterminate, label, forwardedRef, style, ...props } = this.props;
```
You'll probably not want to destructure this as it's needed for the `HiddenCheckbox` component. Probably just want to use `this.props.disabled` on the `StyledCheckbox`.

chanceaclark commented on packages/big-design/src/mixins/index.ts:
```diff
@@ -1,3 +1,4 @@
 export * from './display';
 export * from './margins';
 export * from './paddings';
+export * from './transitions';
```
I don't think we'll want to expose this publicly for now. After you remove it and update the imports, we'll be good to go!

capsula4 replied:
hopefully is all good now

chanceaclark commented:
Fix tests, but should be good to go.
